src/Cookie.php

```diff
@@ -40,88 +71,120 @@ final class Cookie
     private string $value;
 
     /**
-     * @var string|null domain of the cookie
+     * @var DateTimeInterface|null RFC-1123 date at which the cookie expires.
+     * @see https://tools.ietf.org/html/rfc6265#section-4.1.1
      */
-    private ?string $domain = null;
+    private ?DateTimeInterface $expire = null;
 
     /**
-     * @var int the timestamp at which the cookie expires. This is the server timestamp.
-     * Defaults to 0, meaning "until the browser is closed".
+     * @var int|null maximum age of cookie in seconds
      */
-    private int $expire = 0;
+    private ?int $maxAge = null;
 
     /**
-     * @var string the path on the server in which the cookie will be available on. The default is '/'.
+     * @var string|null domain of the cookie.
      */
-    private string $path = '/';
+    private ?string $domain = null;
 
     /**
-     * @var bool whether cookie should be sent via secure connection
+     * @var string|null the path on the server in which the cookie will be available on.
      */
-    private bool $secure = true;
+    private ?string $path = null;
 
     /**
-     * @var bool whether the cookie should be accessible only through the HTTP protocol.
+     * @var bool|null whether cookie should be sent via secure connection
+     */
+    private ?bool $secure = null;
+
+    /**
+     * @var bool|null whether the cookie should be accessible only through the HTTP protocol.
      * By setting this property to true, the cookie will not be accessible by scripting languages,
      * such as JavaScript, which can effectively help to reduce identity theft through XSS attacks.
      */
-    private bool $httpOnly = true;
+    private ?bool $httpOnly = null;
 
     /**
-     * @var string SameSite prevents the browser from sending this cookie along with cross-site requests.
-     * Please note that this feature is only supported since PHP 7.3.0
-     * For better security, an exception will be thrown if `sameSite` is set while using an unsupported version of PHP.
-     * To use this feature across different PHP versions check the version first. E.g.
-     * ```php
-     * $cookie->sameSite = PHP_VERSION_ID >= 70300 ? yii\web\Cookie::SAME_SITE_LAX : null,
-     * ```
-     * @see https://www.owasp.org/index.php/SameSite for more information about sameSite.
+     * @var string|null SameSite prevents the browser from sending this cookie along with cross-site requests.
+     * @see https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#samesite-cookie-attribute for more information about sameSite.
      */
-    private string $sameSite = self::SAME_SITE_LAX;
+    private ?string $sameSite = null;
 
-    public function __construct(string $name, string $value)
+    public function __construct(string $name, string $value = '', bool $safeDefaults = true)
```

Comment:
What is the reason for allowing unsafe defaults? I think we should always start with safe defaults allowing to explicitly change them to unsafe ones.

Reply:
See [Cookie::fromSetCookieString](https://github.com/yiisoft/yii-web/blob/439db1b2fd5befe9a59e2315dcf2ad468f79f7a7/src/Cookie.php#L262), it creates a cookie without any defaults and later populates it from the string. If we don't have such an option, we would need to create a cookie and then chain several methods to reset the defaults while also creating several objects in the process and deleting them (all `clone` statements) and overload GC unnecessarily during that.

Also it is reasonable to allow creating minimal valid cookie without any defaults without any hassle.

Reply:
@samdark I have another idea here. I'm generally against using flags anywhere (constructor, methods...). So what about if I add all available cookie attributes as constructor parameters with their defaults already set to the ones that we consider safe defaults. This will allow to more correct object construction without any flag and will allow also to construct an object with desired attributes without a need to use method chaining for building your desired object in user code, and those methods will be used when needed to modify (actually getting new object, as the class is immutable). So something like this:

```php
public function __construct(
    string $name,
    string $value = '',
    ?DateTimeInterface $expires = null,
    ?int $maxAge = null,
    ?string $domain = null,
    ?string $path = '/',
    ?bool $secure = true,
    ?string $httpOnly = true,
    ?string $sameSite = self::SAME_SITE_LAX
) {
...
}
```

Reply:
How about making constructor private and adding two static factories: `Cookie::create()` and `Cookie::createUnsafe()`?

Reply:
@samdark Yes, that would be better, I will go that route, but I'm not convinced on naming second factory `Cookie::createUnsafe()` as technically it is just some proposed/recommended defaults and not only about safety. What about naming the second one `Cookie::createWithAttributes` or `Cookie::createWithValues` and it will accept all arguments (cookie attributes) as seen in my constructor example and users who just need custom cookie will use that.

Reply:
Your idea about constructor-level defaults looks good. Let's have it.